src/serializers/encoder/with_server_events/v7/encode_graph_entities.c

```diff
@@ -0,0 +1,269 @@
+	for(uint64_t i = 0; i < nodes_to_encode; i++) {
+		Entity *e = (Entity *)DataBlockIterator_Next(iter);
+		// Save ID
+		RedisModule_SaveUnsigned(rdb, e->id);
+		int l = Graph_GetNodeLabel(gc->g, e->id);
+
+		// #labels, currently only one label per node.
+		int label_count = (l == GRAPH_NO_LABEL) ? 0 : 1;
+		RedisModule_SaveUnsigned(rdb, label_count);
+
+		// (label)
+		if(label_count) RedisModule_SaveUnsigned(rdb, l);
+
+		// properties N
+		// (name, value type, value) X N
+		_RdbSaveEntity(rdb, e);
+
+	}
+
+	// Check if done encodeing nodes.
+	if(encoded_nodes + nodes_to_encode == graph_nodes) {
+		DataBlockIterator_Free(iter);
+		iter = NULL;
+	}
+
+	// Update context.
+	GraphEncodeContext_SetDatablockIterator(gc->encoding_context, iter);
+	GraphEncodeContext_SetProcessedNodes(gc->encoding_context, encoded_nodes + nodes_to_encode);
+	_UpdateEncodePhase(gc);
+}
+
+void RdbSaveDeletedEdges_v7(RedisModuleIO *rdb, GraphContext *gc) {
+	/* Format:
+	 * #deleted edges N
+	 * edge id X N */
+
+	// Get graph's deleted edge count.
+	uint64_t deleted_edges = array_len(gc->g->edges->deletedIdx);
+	// Get the number of deleted edges already encoded.
+	uint64_t encoded_deleted_edges = GraphEncodeContext_GetProccessedDeletedEdges(gc->encoding_context);
+	// Calculate the number of deleted edges required to encode in this phase.
+	uint64_t deleted_edges_to_encode = deleted_edges - encoded_deleted_edges;
+	// If the required number is bigger than the allowed entities threshold, set it to be the threshold.
+	if(deleted_edges_to_encode > entities_threshold) deleted_edges_to_encode = entities_threshold;
+
+	// # Deleted edges
+	RedisModule_SaveUnsigned(rdb, deleted_edges_to_encode);
+	// Iterated over the required range in the datablock deleted items.
+	for(uint64_t i = encoded_deleted_edges; i < encoded_deleted_edges + deleted_edges_to_encode; i++) {
+		RedisModule_SaveUnsigned(rdb, gc->g->edges->deletedIdx[i]);
+	}
+	GraphEncodeContext_SetProcessedDeletedEdges(gc->encoding_context,
+												encoded_deleted_edges + deleted_edges_to_encode);
+	_UpdateEncodePhase(gc);
+}
+
+void RdbSaveEdges_v7(RedisModuleIO *rdb, GraphContext *gc) {
+	/* Format:
+	 * #edges (N)
+	 * {
+	 *  source node ID
+	 *  destination node ID
+	 *  relation type
+	 * } X N
+	 * edge properties X N */
+
+	// Get graph's edge count.
+	uint64_t graph_edges = Graph_EdgeCount(gc->g);
+	// Get the number of edges already encoded.
+	uint64_t encoded_edges = GraphEncodeContext_GetProccessedEdges(gc->encoding_context);
+	// Calculate the number of edges required to encode in this phase.
+	uint64_t edges_to_encode = graph_edges - encoded_edges;
+	// If the required number is bigger than the allowed entities threshold, set it to be the threshold.
+	if(edges_to_encode > entities_threshold) edges_to_encode = entities_threshold;
+	// Get current relation matrix.
+	uint r = GraphEncodeContex_GetCurrentRelationID(gc->encoding_context);
+	GrB_Matrix M = Graph_GetRelationMatrix(gc->g, r);
```

Comment:
I crash on this call when instantiating a 5.9 server with:
```
redis-server --loadmodule src/redisgraph.so ENTITIES_THRESHOLD 10
```
And having run:
```
python demo/social/social_demo.py --debug
redis-cli save
```
The backtrace is:
```
#3  0x00007ffff7253412 in __GI___assert_fail (
    assertion=0x7ffff6a1c3e0 "g && (relation_idx == GRAPH_NO_RELATION || relation_idx < Graph_RelationTypeCount(g))",
    file=0x7ffff6a1bec8 "/home/tower-linux/dev/redis/graph/src/graph/graph.c", line=1105,
    function=0x7ffff6a1c890 <__PRETTY_FUNCTION__.13939> "Graph_GetRelationMatrix") at assert.c:101
#4  0x00007ffff642dd88 in Graph_GetRelationMatrix (g=0x5555559a3240, relation_idx=2) at /home/tower-linux/dev/redis/graph/src/graph/graph.c:1105
#5  0x00007ffff644e569 in RdbSaveEdges_v7 (rdb=0x7fffffffb9e0, gc=0x55555599c7f0)
    at /home/tower-linux/dev/redis/graph/src/serializers/encoder/with_server_events/v7/encode_graph_entities.c:209
#6  0x00007ffff644c1df in RdbSaveGraphContext_v7 (rdb=0x7fffffffb9e0, value=0x55555599c7f0)
    at /home/tower-linux/dev/redis/graph/src/serializers/encoder/with_server_events/v7/encode_graphcontext.c:87
#7  0x00007ffff644a498 in RdbSaveGraphContext_WithServerEvents (rdb=0x7fffffffb9e0, value=0x55555599c7f0)
    at /home/tower-linux/dev/redis/graph/src/serializers/encoder/with_server_events/encode_with_server_events.c:11
#8  0x00007ffff64469ba in _GraphContextType_RdbSaveWithServerEvents (rdb=0x7fffffffb9e0, value=0x55555599c7f0)
    at /home/tower-linux/dev/redis/graph/src/serializers/graphcontext_type.c:58
```
I assume that this has to do with the modification of `r` in the loop below, but have not investigated.

Reply:
fixed